### PR TITLE
Adding version string to ApplicationId

### PR DESCRIPTION
**Pull request by gokulavasan**
Build : http://builds.cask.co/browse/CDAP-RUT158

### CODE REVIEW

**Comment by gokulavasan** [comment]
Build is 🍏

File: cdap-proto/src/main/java/co/cask/cdap/proto/id/ApplicationId.java [hunk 1]
```diff
@@ -30,12 +30,18 @@
...
-  public ApplicationId(String namespace, String application) {
+  public ApplicationId(String namespace, String application, String version) {
     super(EntityType.APPLICATION);
     this.namespace = namespace;
     this.application = application;
+    this.version = version;
+  }
+
+  public ApplicationId(String namespace, String application) {
+    this(namespace, application, "-SNAPSHOT");
```
Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/id/ApplicationId.java]: how come its just `-SNAPSHOT`? Shouldn't it have a number before?

File: cdap-proto/src/test/java/co/cask/cdap/proto/audit/AuditMessageTest.java [hunk 2]
```diff
@@ -62,11 +62,12 @@ public void testAccessMessage() throws Exception {
...
       "{\"version\":1,\"time\":2000,\"entityId\":{\"namespace\":\"ns1\",\"stream\":\"stream1\"," +
         "\"entity\":\"STREAM\"},\"user\":\"user1\",\"type\":\"ACCESS\",\"payload\":{\"accessType\":\"WRITE\"," +
-        "\"accessor\":{\"namespace\":\"ns1\",\"application\":\"app1\",\"type\":\"Flow\",\"program\":\"flow1\"," +
-        "\"run\":\"run1\",\"entity\":\"PROGRAM_RUN\"}}}";
+        "\"accessor\":{\"namespace\":\"ns1\",\"application\":\"app1\",\"version\":\"v1\",\"type\":\"Flow\"," +
+        "\"program\":\"flow1\",\"run\":\"run1\",\"entity\":\"PROGRAM_RUN\"}}}";
     AuditMessage flowAccess =
       new AuditMessage(2000L, Ids.namespace("ns1").stream("stream1"), "user1", AuditType.ACCESS,
-                       new AccessPayload(AccessType.WRITE, Ids.namespace("ns1").app("app1").flow("flow1").run("run1")));
+                       new AccessPayload(AccessType.WRITE, Ids.namespace("ns1").app("app1", "v1").flow("flow1")
```
Comment (bdmogal) [inline cdap-proto/src/test/java/co/cask/cdap/proto/audit/AuditMessageTest.java]: nit: unrelated to this PR, but can you replace `Ids.namespace` with `new NamespaceId`?

**Comment by bdmogal** [comment]
@gokulavasan what's the contract of version in the EntityId classes? Is it always going to be a plain string? Shouldn't it be similar to artifacts (major/minor/bugfix)?

File: cdap-proto/src/main/java/co/cask/cdap/proto/id/ApplicationId.java (hunk 1 above)
Comment (gokulavasan) [reply cdap-proto/src/main/java/co/cask/cdap/proto/id/ApplicationId.java]: nope, we are just going to check for -SNAPSHOT suffix to see whether that version is immutable or not. If no version is provided, the default one should be mutable to adhere to the current capability.

**Comment by gokulavasan** [comment]
@bdmogal No, it is going to be a plain string. Refer to the design doc here : https://wiki.cask.co/display/CE/Application+Versioning+Design

**Comment by gokulavasan** [comment]
@chtyim Please review when you get a chance. Addressed comments.

**Comment by gokulavasan** [comment]
@chtyim Modified the EntityId constructors to take in ApplicationId. Please take a look again. Thanks!

File: cdap-proto/src/main/java/co/cask/cdap/proto/id/ApplicationId.java [hunk 3]
```diff
@@ -46,37 +52,45 @@ public String getApplication() {
...
     return new NamespaceId(namespace);
   }
 
+  public ApplicationId version(String version) {
+    return new ApplicationId(namespace, application, version);
+  }
+
   public ProgramId program(ProgramType type, String program) {
-    return new ProgramId(namespace, application, type, program);
+    return new ProgramId(new ApplicationId(namespace, application).version(version), type, program);
```
Comment (chtyim) [inline cdap-proto/src/main/java/co/cask/cdap/proto/id/ApplicationId.java]: Just pass in `this` instead of creating a new `ApplicationId`.

**Comment by gokulavasan** [comment]
@chtyim Addressed comment. Please take a look again. Thank you!

**Comment by chtyim** [comment]
LGTM

**Pull request merged**